### PR TITLE
Adds a confirmation on window exit

### PR DESCRIPTION
**Pull request by edas**
If there are changes not saved, this PR will adds a popup asking for confirmation before leaving the window.

### CODE REVIEW

File: src/hooks/useCollaborationState.js [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    if (dirtySince.current) {
+      dirtySince.current = null
+      setChanges(iterations++)
+    }
+  }
+
+  // Sets dirtySince when the queue fills up
+  function onEnqueueSteps() {
+    if (!dirtySince.current) {
+      if (!dirtySince.current) dirtySince.current = new Date()
```
Comment (y-lohse) [inline src/hooks/useCollaborationState.js]: Same condition twice here

File: src/hooks/useCollaborationState.js [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  const dirtySince = useRef(null)
+  const lastSave = useRef(null)
+  // the state is only here to trigger a rerender when needed
+  const [, setChanges] = useState(iterations)
+
+  // init
+  if (channel) {
+    if (channel.isDirty()) {
+      // always mark as dirty if the channel is dirty
+      if (!dirtySince.current) dirtySince.current = new Date()
```
Comment (y-lohse) [inline src/hooks/useCollaborationState.js]: Can `dirtySince.current` even be set at this stage?

File: src/hooks/useCollaborationState.spec.js [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        const { result } = await initAndQueue()
+        expect(result.current.isDirty).toBeTruthy()
+        await resolvePromise()
+      })
+
+      it('should have a new dirtySince date', async () => {
+        const { result, previous } = await initAndQueue()
+        expect(result.current.dirtySince).not.toBe(previous.dirtySince)
+        await resolvePromise()
+      })
```
Comment (y-lohse) [inline src/hooks/useCollaborationState.spec.js]: 👍 nice comprehensive test suite

File: src/hooks/useCollaborationState.js (hunk 1 above)
Comment (edas) [reply src/hooks/useCollaborationState.js]: Vu et corrigé, merci

File: src/hooks/useCollaborationState.js (hunk 2 above)
Comment (edas) [reply src/hooks/useCollaborationState.js]: Most probably, you always start with a non dirty channel. I prefer to be defensive a prepare the case where the channel is dirty the first time I am executing this hook.

However, this specific line may well be executed even after initialization. If proseMirror enqueues a step, we'll fire an event. This event has to be catched. The event listener will then change de ref value and change the state to call a rerender.

All this is async. We may well have a rerender of the same component at any time, before all this, after all this, or between any step in the chain. So yes, we could have a dirty channel but still not have changed the ref value

File: package.json [hunk 4]
```diff
@@ -50,20 +52,23 @@
     "@atlaskit/media-core": "^31.0.1",
     "@atlaskit/media-editor": "^37.0.3",
     "@atlaskit/smart-card": "^12.6.2",
+    "@material-ui/core": "^4.8.3",
```
[the remark this answers is not in the document]
Comment (edas) [reply package.json]: fixed

File: src/hooks/useCollabProvider.js [hunk 5]
```diff
@@ -9,18 +9,21 @@ function useCollabProvider({ docVersion, noteId, serviceClient }) {
...
+          // The following `collabProviderPlugin` object is defined
+          // in an Atlassian API. The attribute `provider` is expected
+          // to be a Promise, even if we don't need one ourselves here.
+          collabProviderPlugin: {
+            useNativePlugin: true,
+            provider: Promise.resolve(provider),
+            inviteToEditHandler: () => undefined,
+            isInviteToEditButtonSelected: false,
+            userId: serviceClient.getSessionId()
+          }
```
[the remark this answers is not in the document]
Comment (edas) [reply src/hooks/useCollabProvider.js]: This part didn't really changed. I just return both the provider and the plugin where we returned only the plugin before the change.

**Pull request merged**